oioioi/mp/controllers.py

```diff
@@ -0,0 +1,198 @@
+
+CONTEST_RANKING_KEY = 'c'
+
+
+class MPRegistrationController(ParticipantsController):
+    registration_template = 'mp/registration.html'
+
+    @property
+    def form_class(self):
+        from oioioi.mp.forms import MPRegistrationForm
+
+        return MPRegistrationForm
+
+    @property
+    def participant_admin(self):
+        from oioioi.mp.admin import MPRegistrationParticipantAdmin
+
+        return MPRegistrationParticipantAdmin
+
+    @classmethod
+    def anonymous_can_enter_contest(self):
+        return True
+
+    def allow_login_as_public_name(self):
+        return True
+
+    # Redundant because of filter_visible_contests, but saves a db query
+    def can_enter_contest(self, request):
+        return True
+
+    def visible_contests_query(self, request):
+        return Q_always_true()
+
+    def can_register(self, request):
+        return True
+
+    def registration_view(self, request):
+        participant = self._get_participant_for_form(request)
+
+        if 'mp_mpregistrationformdata' in request.session:
+            # pylint: disable=not-callable
+            form = self.form_class(request.session['mp_mpregistrationformdata'])
+            del request.session['mp_mpregistrationformdata']
+        else:
+            form = self.get_form(request, participant)
+        form.set_terms_accepted_text(self.get_terms_accepted_phrase())
+
+        if request.method == 'POST':
+            # pylint: disable=maybe-no-member
+            if form.is_valid():
+                participant, created = Participant.objects.get_or_create(
+                    contest=self.contest, user=request.user
+                )
+                self.handle_validated_form(request, form, participant)
+                if 'next' in request.GET:
+                    return safe_redirect(request, request.GET['next'])
+                else:
+                    return redirect('default_contest_view', contest_id=self.contest.id)
+        can_unregister = False
+        if participant:
+            can_unregister = self.can_unregister(request, participant)
+        context = {
+            'form': form,
+            'participant': participant,
+            'can_unregister': can_unregister,
+            'contest_name': self.contest.name
+        }
+        return TemplateResponse(request, self.registration_template, context)
+
+    def mixins_for_admin(self):
+        from oioioi.participants.admin import TermsAcceptedPhraseAdminMixin
+
+        return super(MPRegistrationController, self).mixins_for_admin() + (
+            TermsAcceptedPhraseAdminMixin,
+        )
+
+    def can_change_terms_accepted_phrase(self, request):
+        return not MPRegistration.objects.filter(
+            participant__contest=request.contest
+        ).exists()
+
+
+class MPContestController(ProgrammingContestController):
+    description = _("Mistrz Programowania")
+    create_forum = False
+
+    show_email_in_participants_data = True
```

Comment:
I will have to make sure we can do that from legal point of view.

Reply:
It turned out that the organizers already signed data processing agreements and can access participants personal data.